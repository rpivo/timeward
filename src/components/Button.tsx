import React from 'react';
import StyledButton from '@styles/components/Button.styled';
import buttonPaths from '@utilities/svg';

export type ButtonProps = {
  kind: string;
  handleClick: () => void;
}

const Button: React.FC<ButtonProps> = ({ kind, handleClick }: ButtonProps): JSX.Element => {

  const Start: React.FC = (): JSX.Element =>
    <svg width="15" height="15" viewBox="0 0 50 50">
      <path d={ buttonPaths.start } transform="translate(54.392 -7.309) rotate(90)" fill="#aaa" />
    </svg>;

  const Pause: React.FC = (): JSX.Element =>
    <svg width="15" height="15" viewBox="0 0 50 50">
      <g transform="translate(33 404)">
        <path d={ buttonPaths.pause } transform="translate(-5.034 -404)" fill="#aaa"/>
        <path d={ buttonPaths.pause } transform="translate(-33 -404)" fill="#aaa"/>
      </g>
    </svg>;

  const Stop: React.FC = (): JSX.Element =>
    <svg width="15" height="15" viewBox="0 0 50 50">
      <path d={ buttonPaths.stop } fill="#aaa"/>
    </svg>;

  const renderSwitch = (buttonType: string): JSX.Element => {
    switch (buttonType) {
    case 'start':
      return <Start />;
    case 'pause':
      return <Pause />;
    default:
      return <Stop />;
    }
  };

  return (
    <StyledButton kind={ kind } onClick={ handleClick }>
      { renderSwitch(kind) }
    </StyledButton>
  );
};

export default Button;
