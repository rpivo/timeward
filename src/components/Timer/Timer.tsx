import * as React from 'react';

export const Timer = (): JSX.Element => <div></div>;