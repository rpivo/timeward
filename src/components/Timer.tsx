import React from 'react';
import { Button } from '@components/Button';

interface TimerState { time: string }

export class Timer extends React.Component <{}, TimerState> {
  public state: TimerState = { time: '00:00:00' };

  public render(): JSX.Element {
    return (
      <div>
        { this.state.time }
        <Button kind='test' />
      </div>
    );
  }
}
