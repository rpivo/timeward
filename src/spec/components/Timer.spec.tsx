import React from 'react';
import renderer from 'react-test-renderer';
import { mount } from 'enzyme';
import { Timer } from '@components/Timer';
import { Button } from '@components/Button';

describe('Timer', () => {
  interface TimerInstance extends React.Component<{}, {}> {
    seconds: number;
    minutes: number;
    hours: number;
  }

  const wrapper = mount<Timer>(<Timer />);
  const timer = wrapper.find(Timer);
  const TimerInstance = timer.instance() as TimerInstance;

  beforeEach(() => {
    wrapper.state().seconds = 0;
    TimerInstance.hours = 0;
    TimerInstance.minutes = 0;
    TimerInstance.seconds = 0;
  });

  describe('render', () => {
    const testRenderer = renderer.create(<Timer />);
    const testInstance = testRenderer.root;

    it('should render correctly', () => {
      testRenderer.toJSON();
      expect(testRenderer).toMatchSnapshot();
    });

    it('should contain two Button components', () => {
      const children = testInstance.findAllByType(Button);
      expect(children.length).toEqual(2);
    });
  });

  describe('state', () => {
    expect(wrapper.state('buttonType')).toBe('start');
    expect(wrapper.state('seconds')).toBe(0);
  });

  describe('methods', () => {
    const instance = wrapper.instance();

    describe('constructStringFromSeconds', () => {
      it('should return a timestring', () => {
        expect(instance['constructStringFromSeconds']()).toBe('00:00:00');
      });

      it('should set this.seconds to 0 under certain conditions', () => {
        wrapper.state().seconds = 60;
        instance['constructStringFromSeconds']();
        expect(TimerInstance.seconds).toBe(0);
      });

      it('should set this.minutes to 0 under certain conditions', () => {
        wrapper.state().seconds = 60;
        TimerInstance.minutes = 59;
        instance['constructStringFromSeconds']();
        expect(TimerInstance.hours).toBe(1);
      });
    });

    it('should call startTimer', () => {
      instance['startTimer']();

      expect(wrapper.state('buttonType')).toBe('pause');
      expect(wrapper.state('seconds')).toBe(0);
    });

    it('should call pauseTimer', () => {
      instance['pauseTimer']();

      expect(wrapper.state('buttonType')).toBe('start');
    });

    it('should call stopTimer', () => {
      instance['stopTimer']();

      expect(wrapper.state('buttonType')).toBe('start');
      expect(wrapper.state('seconds')).toBe(0);
    });
  });
});
