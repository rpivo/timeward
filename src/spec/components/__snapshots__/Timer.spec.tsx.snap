// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Timer render should render correctly 1`] = `
<div
  className="sc-AykKD cstrHd"
>
  <span>
    00:00:00
    
  </span>
  <div
    className="sc-AykKC cSvOKE"
    kind="start"
    onClick={[Function]}
  >
    <svg
      height="15"
      viewBox="0 0 50 50"
      width="15"
    >
      <path
        d="M28.682,12.087a4.188,4.188,0,0,1,7.254,0L56.741,48.109a4.188,4.188,0,0,1-3.627,6.283H11.5a4.188,4.188,0,0,1-3.627-6.283Z"
        fill="#aaa"
        transform="translate(54.392 -7.309) rotate(90)"
      />
    </svg>
  </div>
  <div
    className="sc-AykKC cSvOKE"
    kind="stop"
    onClick={[Function]}
  >
    <svg
      height="15"
      viewBox="0 0 50 50"
      width="15"
    >
      <path
        d="M4.237,0H45.763A4.237,4.237,0,0,1,50,4.237V45.763A4.237,4.237,0,0,1,45.763,50H4.237A4.237,4.237,0,0,1,0,45.763V4.237A4.237,4.237,0,0,1,4.237,0Z"
        fill="#aaa"
      />
    </svg>
  </div>
</div>
`;
